c10/util/Exception.h

```diff
@@ -112,17 +113,65 @@ class C10_API Error : public std::exception {
   std::string compute_what(bool include_backtrace) const;
 };
 
+class C10_API Warning {
+ public:
+  class C10_API UserWarning {};
+  using warning_variant_t = c10::variant<UserWarning>;
```

Comment:
It is probably a good idea to add another warning type, even if we don't exhaustively go through and use it everywhere we should at this point. Having two may uncover some things.

Reply:
I can add `c10::DeprecationWarning` and use it in one place, with a unit test to make sure the Python translation works.

Just to be clear, this will require adding `TORCH_WARN_DEPRECATION` and `TORCH_WARN_WITH`

Reply:
I added `TORCH_WARN_DEPRECATION`, used it in `torch.ger`, and added a unit test for it